contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,850 @@
+ the Pod. In place resource reallocation is out of scope for this design. 
+ Vertical scaling must be performed destructively.
+ - The primary means of configuration update will be configuration files, 
+ command line flags, environment variables, or ConfigMaps consumed as the one 
+ of the former. 
+ - In place configuration update via SIGHUP is not universally 
+ supported, and Kubelet provides no mechanism to perform this currently. Pod 
+ reconfiguration will be performed destructively.
+ - Stateful applications are likely to evolve wire protocols and storage formats
+  between versions. In most cases, when updating the application's Pod's 
+  containers, it will not be safe to roll back or forward to an arbitrary 
+  version. StatefulSet update should work well when rolling out an update, 
+  or performing a rollback, between two specific revisions of the StatefulSet.
+
+## Requirements
+This design is based on the following requirements.
+- Users must be able to update the containers of a StatefulSet's Pods.
+  - Updates to container commands, images, resources and configuration must be 
+  supported.
+- The update must progress in a sequential, deterministic order and respect the 
+  StatefulSet
+  [identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity), 
+  [deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+  guarantees.
+- A failed update must halt.
+- Users must be able to rollback an update.
+- Users must be able to roll forward to fix a failing/failed update.
+- Users must be able to view the status of an update.
+- Users should be able to view a bounded history of the updates that have been 
+applied to the StatefulSet.
+
+## API Object
+
+The following modifications will be made to the StatefulSetStatus API object.
+
+```go
+ type StatefulSetStatus struct {
+    // ObservedGeneration and Replicas fields are ommitted for brevity.
+
+ 	// CurrentTemplateRevision, if not nil, is the revision of the PodTemplate
+ 	// that was used to create Pods with ordinals in the sequence
+    // [0,CurrentRevisionReplicas).
+ 	TemplateRevision  *int64 `json:"templateRevision,omitempty"`
+ 	
+ 	// TargetTemplateRevision, if not nil, is the revision of the PodTemplate 
+ 	// that was used to create Pods with ordinals in the sequence 
+    // [Replicas - TargetRevisionReplicas, Replicas).
+ 	TargetTemplateRevision *int64 `json:"targetTemplateRevision,omitempty"`
+ 	
+ 	// ReadyReplicas is the current number of Pods, created by the StatefulSet
+    // controller, that have a Status for Running and a Ready Condition.
+ 	ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+ 	
+ 	// CurrentRevisionReplicas is the number of Pods created by the StatefulSet 
+    // controller from the PodTemplateSpec indicated by CurrentTemplateRevision.
+ 	CurrentReplicas int32 `json:"currentReplicas,omitempty"`
```

Comment:
This may be confusing because once a StatefulSet is updated and a new PodTemplate is created, CurrentReplicas is really NeedToBeReplacedReplicas. I can't think of a better name now.